tests/bsim_bt/bsim_test_app/src/test_connect1.c

```diff
@@ -0,0 +1,368 @@
+{
+	bt_addr_le_t *addr = user_data;
+	int i;
+
+	printk("[AD]: %u data_len %u\n", type, data_len);
+
+	switch (type) {
+	case BT_DATA_UUID16_SOME:
+	case BT_DATA_UUID16_ALL:
+		if (data_len % sizeof(u16_t) != 0) {
+			printk("AD malformed\n");
+			return true;
+		}
+
+		for (i = 0; i < data_len; i += sizeof(u16_t)) {
+			struct bt_uuid *uuid;
+			u16_t u16;
+			int err;
+
+			memcpy(&u16, &data[i], sizeof(u16));
+			uuid = BT_UUID_DECLARE_16(sys_le16_to_cpu(u16));
+			if (bt_uuid_cmp(uuid, BT_UUID_HRS)) {
+				continue;
+			}
+
+			err = bt_le_scan_stop();
+			if (err) {
+				printk("Stop LE scan failed (err %d)\n", err);
+				continue;
+			}
+
+			default_conn = bt_conn_create_le(addr,
+					BT_LE_CONN_PARAM_DEFAULT);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static void ad_parse(struct net_buf_simple *ad,
+		bool (*func)(u8_t type, const u8_t *data,
+				u8_t data_len, void *user_data),
+				void *user_data)
+{
+	while (ad->len > 1) {
+		u8_t len = net_buf_simple_pull_u8(ad);
+		u8_t type;
+
+		/* Check for early termination */
+		if (len == 0) {
+			return;
+		}
+
+		if (len > ad->len) {
+			printk("AD malformed\n");
+			return;
+		}
+
+		type = net_buf_simple_pull_u8(ad);
+
+		if (!func(type, ad->data, len - 1, user_data)) {
+			return;
+		}
+
+		net_buf_simple_pull(ad, len - 1);
+	}
+}
+
+static void device_found(const bt_addr_le_t *addr, s8_t rssi, u8_t type,
+		struct net_buf_simple *ad)
+{
+	char dev[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, dev, sizeof(dev));
+	printk("[DEVICE]: %s, AD evt type %u, AD data len %u, RSSI %i\n",
+			dev, type, ad->len, rssi);
+
+	/* We're only interested in connectable events */
+	if (type == BT_LE_ADV_IND || type == BT_LE_ADV_DIRECT_IND) {
+		/* TODO: Move this to a place it can be shared */
+		ad_parse(ad, eir_found, (void *)addr);
+	}
+}
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	int err;
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Disconnected: %s (reason %u)\n", addr, reason);
+
+	if (default_conn != conn) {
+		return;
+	}
+
+	bt_conn_unref(default_conn);
+	default_conn = NULL;
+
+	/* This demo doesn't require active scan */
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, device_found);
+	if (err) {
+		printk("Scanning failed to start (err %d)\n", err);
+	}
+}
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+static void test_con1_main(void)
+{
+	int err;
+
+	err = bt_enable(NULL);
+
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	printk("Bluetooth initialized\n");
+
+	bt_conn_cb_register(&conn_callbacks);
+
+	err = bt_le_scan_start(BT_LE_SCAN_ACTIVE, device_found);
+
+	if (err) {
+		printk("Scanning failed to start (err %d)\n", err);
+		return;
+	}
+
+	printk("Scanning successfully started\n");
+}
+
+static const bst_test_instance_t test_unencrypted[] = {
+	{"connect",        /* Test Name */
```

Comment:
Doesn't the coding style recommend using explicit struct member initializers like:
```
{
        .test_id = "connect",
        .test_descr = "Basic connection test.",
        ...
}
```

This is IMO more readable, you don't need the intermediate NULL entries (just leave something out if you don't use it) and it's also more robust since now you don't depend on the ordering of variables inside the struct.

Btw, why is this struct a typedef? AFAIK only types where the API user is not expected to directly access the struct internals should be typedefs (i.e. the meaning a typedef conveys is that the type is considered opaque)

Reply:
Thanks Johan. I agree. Fixed.